earthaccess/api.py

```diff
@@ -162,16 +162,16 @@ def download(
        * If we run it outside AWS (us-west-2 region) and the dataset is cloud hostes we'll use HTTP links
 
     Parameters:
-        granules: a granule, list of granules, or a list of granule links (HTTP)
+        granules: a granule, list of granules, a granule link (HTTP), or a list of granule links (HTTP)
         local_path: local directory to store the remote data granules
         provider: if we download a list of URLs we need to specify the provider.
         threads: parallel number of threads to use to download the files, adjust as necessary, default = 8
 
     Returns:
         List of downloaded files
     """
-    if isinstance(granules, DataGranule):
-        granules = [granules]
+    if isinstance(granules, (DataGranule, str)):
+        granules = cast(Union[List[DataGranule], List[str]], [granules])
```

Comment:
This `cast` is a little messy, but is to make `mypy` happy. I'm not a typing expert -- definitely open to other suggestions that make this less involved.

Reply:
This one feels like it should be easy :laughing: It's such a common pattern for a function to accept one or many of a thing. I'm sure there's a term-of-art of this, but my memory and google skills are failing me.

This method avoids telling the typechecker "trust us" so I think it's a bit better, but it's also still pretty messy. With the combined `isinstance` check, Mypy sees `thing` as `str | Thing`, but then after listing it Mypy `list[str | Thing]` instead of what I expected expected, `list[str] | list[thing]`.

```python
class Thing:
    pass


def foo(things: Thing | str | list[Thing] | list[str]) -> list[str] | list[Thing]:
    thing_list: list[str] | list[Thing]
    if isinstance(things, Thing): 
        thing_list = [things]
    elif isinstance(things, str):
        thing_list = [things]

    return thing_list
```

Reply:
I'm going to share this out and see if anyone wants to play code golf.

Reply:
An entry using a generic function and switch statement: https://gist.github.com/kbeamnsidc/4ffa3708677fa5b36623071443a194c7